compiler/backend/src/org/jetbrains/kotlin/codegen/inline/LambdaInfo.kt

```diff
@@ -129,15 +130,18 @@ abstract class DefaultLambda(
 
     // Returns whether the loaded invoke is erased, i.e. the name equals the fallback and all types are `Object`.
     protected fun loadInvoke(sourceCompiler: SourceCompilerForInline, erasedName: String, actualMethod: Method): Boolean {
-        val classBytes = loadClass(sourceCompiler)
-        // TODO: `signatureAmbiguity = true` ignores the argument types from `invokeMethod` and only looks at the count.
-        node = getMethodNode(classBytes, actualMethod.name, actualMethod.descriptor, lambdaClassType, signatureAmbiguity = true)
-            ?: getMethodNode(classBytes, erasedName, actualMethod.descriptor, lambdaClassType, signatureAmbiguity = true)
-                    ?: error("Can't find method '$actualMethod' in '${lambdaClassType.internalName}'")
+        node = getMethodNodeImprecise(loadClass(sourceCompiler), lambdaClassType, actualMethod, erasedName)
+            ?: error("Can't find method '$actualMethod' in '${lambdaClassType.internalName}'")
         return invokeMethod.run { name == erasedName && returnType == OBJECT_TYPE && argumentTypes.all { it == OBJECT_TYPE } }
     }
 
     private companion object {
+        fun getMethodNodeImprecise(classBytes: ByteArray, classType: Type, method: Method, erasedName: String) =
+            getMethodNode(classBytes, classType) { it, access ->
```

Comment:
If both method would be present in class file then order of methods would make sense, wouldn't it?

Reply:
Or `access.and(Opcodes.ACC_SYNTHETIC)` check allows to avoid ambiguity?

Reply:
There should be only one non-synthetic `invoke` (the one that contains the real code written by the user), yes.